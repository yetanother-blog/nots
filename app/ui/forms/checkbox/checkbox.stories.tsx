import type { Meta, StoryObj } from '@storybook/react';
import { Checkbox } from './checkbox';
import { action } from '@storybook/addon-actions';
import { useState } from 'react';
import { Button } from '~/ui';

const meta: Meta<typeof Checkbox> = {
  title: 'Forms/Checkbox',
  component: Checkbox,
};

export default meta;
type Story = StoryObj<typeof Checkbox>;

/**
 * Checkbox allow users to accept or select something in forms (e.g. terms and
 * conditions).
 */
export const CheckboxStates: Story = {
  render: () => (
    <div className="flex flex-col gap-6 dark:bg-nots-grey-800">
      <Checkbox label="Default" id="default" onChange={action('changed')} />
      <Checkbox
        label="Checked"
        id="default-checked"
        checked={true}
        onChange={action('changed')}
      />
      <Checkbox
        label="Default w/ helper text"
        id="helper-text"
        value="This is a text field with a helper text"
        helperText="This is a helper text"
        onChange={action('changed')}
      />
      <Checkbox
        label="Disabled"
        id="disabled-text-field"
        disabled={true}
        onChange={action('changed')}
      />
      <Checkbox
        label="Checked and disabled"
        id="disabled-text-field"
        disabled={true}
        checked={true}
        onChange={action('changed')}
      />
      <Checkbox
        label="With error message"
        id="error-text-field"
        isError={true}
        helperText="Something went wrong"
        onChange={action('changed')}
      />
      <Checkbox
        label="With long text: Minim qui reprehenderit cillum duis et commodo occaecat officia labore dolore consectetur. Lorem cupidatat tempor irure voluptate laboris non magna tempor aute fugiat ipsum. Ipsum consectetur id reprehenderit dolore id aliqua laboris eu ipsum culpa non."
        id="with-long-text"
        checked={true}
        onChange={action('changed')}
      />
    </div>
  ),
};

export const CheckboxIndeterminate: Story = {
  render: () => {
    // eslint-disable-next-line react-hooks/rules-of-hooks
    const [indeterminate, setIndeterminate] = useState(false);

    // eslint-disable-next-line react-hooks/rules-of-hooks
    const [checked, setChecked] = useState(false);

    const toggleIndeterminate = () => {
      setIndeterminate((i) => !i);
    };

    const toggleChecked = () => {
      setChecked((c) => !c);
      setIndeterminate(false);
    };

    return (
      <div className="flex gap-4 items-center">
        <Checkbox
          label="Indeterminate"
          id="indeterminate"
          indeterminate={indeterminate}
          checked={checked}
          onChange={toggleChecked}
        />
        <Button onClick={toggleIndeterminate}>Toggle State</Button>
      </div>
    );
  },
};
