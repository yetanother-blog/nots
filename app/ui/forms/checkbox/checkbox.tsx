import type { InputHTMLAttributes, ReactNode, Ref } from 'react';
import { createRef, forwardRef, useEffect } from 'react';
import { Label } from '../label/label';
import { FormWrapper } from '../form-wrapper/form-wrapper';
import { FormFieldStyle } from '../form-field-style/form-field-style';
import { clsx } from 'clsx';

export type CheckboxProps = {
  label?: ReactNode;
  isError?: boolean;
  helperText?: ReactNode;
  indeterminate?: boolean;
} & Omit<InputHTMLAttributes<HTMLInputElement>, 'type'>;

export const Checkbox = forwardRef(
  (
    {
      label,
      isError,
      indeterminate = false,
      helperText,
      ...inputProps
    }: CheckboxProps,
    ref: Ref<HTMLInputElement>
  ) => {
    const inputRef = ref ?? createRef<HTMLInputElement>();

    useEffect(() => {
      if (!inputRef || inputRef instanceof Function || !inputRef.current) {
        return;
      }

      inputRef.current.indeterminate = indeterminate;
    }, [indeterminate, inputRef]);

    const interdeterminateClasses =
      'indeterminate:after:bg-indeterminate indeterminate:border-nots-grey-800 indeterminate:bg-nots-grey-800 indeterminate:after:block indeterminate:hover:border-nots-primary-700 indeterminate:hover:nots-ws-primary-700 indeterminate:after:h-[24px] indeterminate:after:w-[24px]';

    const interdeterminateClassesDark =
      'dark:indeterminate:after:invert dark:indeterminate:border-nots-grey-50 dark:indeterminate:bg-white dark:indeterminate:hover:border-nots-grey-100 dark:indeterminate:hover:bg-nots-grey-100';

    const checkedClasses =
      'checked:after:bg-checkmark checked:border-nots-grey-800 checked:bg-nots-grey-800 checked:after:block checked:hover:border-nots-grey-600 checked:hover:bg-nots-grey-600 checked:after:h-[13px] checked:after:w-[13px]';

    const checkedClassesDark =
      'dark:checked:border-nots-grey-50 dark:checked:bg-white dark:checked:after:invert dark:checked:hover:border-nots-grey-100 dark:checked:hover:bg-nots-grey-100';

    const disabledClasses =
      'disabled:border-nots-grey-100 disabled:bg-nots-grey-50 disabled:checked:border-nots-grey-700 disabled:checked:bg-nots-grey-700';

    const disabledClassesDark =
      'dark:disabled:border-nots-grey-700 dark:disabled:bg-nots-grey-700 dark:disabled:checked:border-nots-grey-700 dark:disabled:checked:bg-nots-grey-200 dark:disabled:checked:after:opacity-30';

    const classes = clsx(
      `peer grid h-7 w-7 shrink-0 appearance-none cursor-pointer place-content-center after:hidden after:content-['']`,
      interdeterminateClasses,
      interdeterminateClassesDark,
      checkedClasses,
      checkedClassesDark,
      disabledClasses,
      disabledClassesDark
    );

    return (
      <FormWrapper isError={isError} helperText={helperText}>
        <div className="flex gap-2">
          <FormFieldStyle
            as="input"
            isError={isError}
            className={classes}
            ref={inputRef}
            type="checkbox"
            {...inputProps}
          />
          {label && (
            <Label
              htmlFor={inputProps.id}
              className="cursor-pointer pt-[4px] peer-disabled:pointer-events-none"
            >
              {label}
            </Label>
          )}
        </div>
      </FormWrapper>
    );
  }
);

Checkbox.displayName = 'Checkbox';
