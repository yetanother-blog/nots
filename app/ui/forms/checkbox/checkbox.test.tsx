import { render } from '@testing-library/react';
import { createRef } from 'react';
import { Checkbox } from './checkbox';

describe('Checkbox', () => {
  it('should render unticked checkbox', () => {
    const { container } = render(
      <Checkbox label="The default" id="default" onChange={() => {}} />
    );
    expect(container).toMatchSnapshot();
  });

  it('should render ticked checkbox', () => {
    const { container } = render(
      <Checkbox
        label="The default"
        id="default"
        onChange={() => {}}
        checked={true}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it('should render disabled and unticked checkbox', () => {
    const { container } = render(
      <Checkbox
        label="The disabled text field"
        id="disabled"
        disabled={true}
        value="This is a disabled text field"
        onChange={() => {}}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it('should render disabled and ticked checkbox', () => {
    const { container } = render(
      <Checkbox
        label="The disabled text field"
        id="disabled"
        disabled={true}
        checked={true}
        value="This is a disabled text field"
        onChange={() => {}}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it('should render checkbox w/ helper text', () => {
    const { container } = render(
      <Checkbox
        label="The helper text"
        id="helper"
        helperText="This is a helper text"
        onChange={() => {}}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it('should render checkbox w/ error', () => {
    const { container } = render(
      <Checkbox
        label="The error text field"
        id="error"
        isError={true}
        helperText="This is an error text field"
        onChange={() => {}}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it('should wire up ref correctly', () => {
    const ref = createRef<HTMLInputElement>();
    render(<Checkbox id="default" ref={ref} />);

    expect(ref.current).toMatchSnapshot();
  });

  it('should set checkbox to indeterminate', () => {
    const { container } = render(
      <Checkbox id="indeterminate" indeterminate={true} />
    );

    const checkbox = container.querySelector('input');

    expect(checkbox?.indeterminate).toBe(true);
  });

  it('should unset indeterminate state', () => {
    const { queryByRole, rerender } = render(
      <Checkbox id="indeterminate" indeterminate={true} />
    );

    rerender(<Checkbox id="indeterminate" indeterminate={false} />);

    const checkbox = queryByRole('checkbox') as HTMLInputElement;
    expect(checkbox.indeterminate).toBe(false);
  });
});
