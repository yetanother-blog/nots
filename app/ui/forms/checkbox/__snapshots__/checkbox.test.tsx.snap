// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Checkbox > should render checkbox w/ error 1`] = `
<div>
  <div
    class="flex flex-col gap-2"
  >
    <div
      class="flex gap-2"
    >
      <input
        class="rounded-lg border transition-all focus-within:outline-none focus-within:ring-4 active:ring-4 border-ws-danger-500  ring-ws-danger-300 focus-within:border-ws-danger-500  hover:border-ws-danger-500 peer grid h-7 w-7 shrink-0 appearance-none cursor-pointer place-content-center after:hidden after:content-[''] indeterminate:after:bg-indeterminate indeterminate:border-ws-primary-500 indeterminate:bg-ws-primary-500 indeterminate:after:block indeterminate:hover:border-ws-primary-700 indeterminate:hover:bg-ws-primary-700 indeterminate:after:h-[24px] indeterminate:after:w-[24px] checked:after:bg-checkmark checked:border-ws-primary-500 checked:bg-ws-primary-500 checked:after:block checked:hover:border-ws-primary-700 checked:hover:bg-ws-primary-700 checked:after:h-[13px] checked:after:w-[13px] disabled:border-ws-grey-100 disabled:bg-ws-grey-50 disabled:checked:border-ws-primary-300 disabled:checked:bg-ws-primary-300"
        id="error"
        type="checkbox"
      />
      <label
        class="antialiased text-sm text-ws-grey-600 font-semibold font-basier-circle cursor-pointer pt-[4px] peer-disabled:pointer-events-none"
        for="error"
      >
        The error text field
      </label>
    </div>
    <p
      class="antialiased text-xs font-normal text-ws-danger-500 font-basier-circle"
    >
      This is an error text field
    </p>
  </div>
</div>
`;

exports[`Checkbox > should render checkbox w/ helper text 1`] = `
<div>
  <div
    class="flex flex-col gap-2"
  >
    <div
      class="flex gap-2"
    >
      <input
        class="rounded-lg border transition-all focus-within:outline-none focus-within:ring-4 active:ring-4 border-ws-grey-300  ring-ws-primary-300 focus-within:border-ws-primary-500  hover:border-ws-primary-500 peer grid h-7 w-7 shrink-0 appearance-none cursor-pointer place-content-center after:hidden after:content-[''] indeterminate:after:bg-indeterminate indeterminate:border-ws-primary-500 indeterminate:bg-ws-primary-500 indeterminate:after:block indeterminate:hover:border-ws-primary-700 indeterminate:hover:bg-ws-primary-700 indeterminate:after:h-[24px] indeterminate:after:w-[24px] checked:after:bg-checkmark checked:border-ws-primary-500 checked:bg-ws-primary-500 checked:after:block checked:hover:border-ws-primary-700 checked:hover:bg-ws-primary-700 checked:after:h-[13px] checked:after:w-[13px] disabled:border-ws-grey-100 disabled:bg-ws-grey-50 disabled:checked:border-ws-primary-300 disabled:checked:bg-ws-primary-300"
        id="helper"
        type="checkbox"
      />
      <label
        class="antialiased text-sm text-ws-grey-600 font-semibold font-basier-circle cursor-pointer pt-[4px] peer-disabled:pointer-events-none"
        for="helper"
      >
        The helper text
      </label>
    </div>
    <p
      class="antialiased text-xs font-normal text-ws-grey-600 font-basier-circle"
    >
      This is a helper text
    </p>
  </div>
</div>
`;

exports[`Checkbox > should render disabled and ticked checkbox 1`] = `
<div>
  <div
    class="flex flex-col gap-2"
  >
    <div
      class="flex gap-2"
    >
      <input
        class="rounded-lg border transition-all focus-within:outline-none focus-within:ring-4 active:ring-4 border-ws-grey-300  ring-ws-primary-300 focus-within:border-ws-primary-500  hover:border-ws-primary-500 pointer-events-none bg-ws-grey-50 peer grid h-7 w-7 shrink-0 appearance-none cursor-pointer place-content-center after:hidden after:content-[''] indeterminate:after:bg-indeterminate indeterminate:border-ws-primary-500 indeterminate:bg-ws-primary-500 indeterminate:after:block indeterminate:hover:border-ws-primary-700 indeterminate:hover:bg-ws-primary-700 indeterminate:after:h-[24px] indeterminate:after:w-[24px] checked:after:bg-checkmark checked:border-ws-primary-500 checked:bg-ws-primary-500 checked:after:block checked:hover:border-ws-primary-700 checked:hover:bg-ws-primary-700 checked:after:h-[13px] checked:after:w-[13px] disabled:border-ws-grey-100 disabled:bg-ws-grey-50 disabled:checked:border-ws-primary-300 disabled:checked:bg-ws-primary-300"
        disabled=""
        id="disabled"
        type="checkbox"
        value="This is a disabled text field"
      />
      <label
        class="antialiased text-sm text-ws-grey-600 font-semibold font-basier-circle cursor-pointer pt-[4px] peer-disabled:pointer-events-none"
        for="disabled"
      >
        The disabled text field
      </label>
    </div>
  </div>
</div>
`;

exports[`Checkbox > should render disabled and unticked checkbox 1`] = `
<div>
  <div
    class="flex flex-col gap-2"
  >
    <div
      class="flex gap-2"
    >
      <input
        class="rounded-lg border transition-all focus-within:outline-none focus-within:ring-4 active:ring-4 border-ws-grey-300  ring-ws-primary-300 focus-within:border-ws-primary-500  hover:border-ws-primary-500 pointer-events-none bg-ws-grey-50 peer grid h-7 w-7 shrink-0 appearance-none cursor-pointer place-content-center after:hidden after:content-[''] indeterminate:after:bg-indeterminate indeterminate:border-ws-primary-500 indeterminate:bg-ws-primary-500 indeterminate:after:block indeterminate:hover:border-ws-primary-700 indeterminate:hover:bg-ws-primary-700 indeterminate:after:h-[24px] indeterminate:after:w-[24px] checked:after:bg-checkmark checked:border-ws-primary-500 checked:bg-ws-primary-500 checked:after:block checked:hover:border-ws-primary-700 checked:hover:bg-ws-primary-700 checked:after:h-[13px] checked:after:w-[13px] disabled:border-ws-grey-100 disabled:bg-ws-grey-50 disabled:checked:border-ws-primary-300 disabled:checked:bg-ws-primary-300"
        disabled=""
        id="disabled"
        type="checkbox"
        value="This is a disabled text field"
      />
      <label
        class="antialiased text-sm text-ws-grey-600 font-semibold font-basier-circle cursor-pointer pt-[4px] peer-disabled:pointer-events-none"
        for="disabled"
      >
        The disabled text field
      </label>
    </div>
  </div>
</div>
`;

exports[`Checkbox > should render ticked checkbox 1`] = `
<div>
  <div
    class="flex flex-col gap-2"
  >
    <div
      class="flex gap-2"
    >
      <input
        class="rounded-lg border transition-all focus-within:outline-none focus-within:ring-4 active:ring-4 border-ws-grey-300  ring-ws-primary-300 focus-within:border-ws-primary-500  hover:border-ws-primary-500 peer grid h-7 w-7 shrink-0 appearance-none cursor-pointer place-content-center after:hidden after:content-[''] indeterminate:after:bg-indeterminate indeterminate:border-ws-primary-500 indeterminate:bg-ws-primary-500 indeterminate:after:block indeterminate:hover:border-ws-primary-700 indeterminate:hover:bg-ws-primary-700 indeterminate:after:h-[24px] indeterminate:after:w-[24px] checked:after:bg-checkmark checked:border-ws-primary-500 checked:bg-ws-primary-500 checked:after:block checked:hover:border-ws-primary-700 checked:hover:bg-ws-primary-700 checked:after:h-[13px] checked:after:w-[13px] disabled:border-ws-grey-100 disabled:bg-ws-grey-50 disabled:checked:border-ws-primary-300 disabled:checked:bg-ws-primary-300"
        id="default"
        type="checkbox"
      />
      <label
        class="antialiased text-sm text-ws-grey-600 font-semibold font-basier-circle cursor-pointer pt-[4px] peer-disabled:pointer-events-none"
        for="default"
      >
        The default
      </label>
    </div>
  </div>
</div>
`;

exports[`Checkbox > should render unticked checkbox 1`] = `
<div>
  <div
    class="flex flex-col gap-2"
  >
    <div
      class="flex gap-2"
    >
      <input
        class="rounded-lg border transition-all focus-within:outline-none focus-within:ring-4 active:ring-4 border-ws-grey-300  ring-ws-primary-300 focus-within:border-ws-primary-500  hover:border-ws-primary-500 peer grid h-7 w-7 shrink-0 appearance-none cursor-pointer place-content-center after:hidden after:content-[''] indeterminate:after:bg-indeterminate indeterminate:border-ws-primary-500 indeterminate:bg-ws-primary-500 indeterminate:after:block indeterminate:hover:border-ws-primary-700 indeterminate:hover:bg-ws-primary-700 indeterminate:after:h-[24px] indeterminate:after:w-[24px] checked:after:bg-checkmark checked:border-ws-primary-500 checked:bg-ws-primary-500 checked:after:block checked:hover:border-ws-primary-700 checked:hover:bg-ws-primary-700 checked:after:h-[13px] checked:after:w-[13px] disabled:border-ws-grey-100 disabled:bg-ws-grey-50 disabled:checked:border-ws-primary-300 disabled:checked:bg-ws-primary-300"
        id="default"
        type="checkbox"
      />
      <label
        class="antialiased text-sm text-ws-grey-600 font-semibold font-basier-circle cursor-pointer pt-[4px] peer-disabled:pointer-events-none"
        for="default"
      >
        The default
      </label>
    </div>
  </div>
</div>
`;

exports[`Checkbox > should wire up ref correctly 1`] = `
<input
  class="rounded-lg border transition-all focus-within:outline-none focus-within:ring-4 active:ring-4 border-ws-grey-300  ring-ws-primary-300 focus-within:border-ws-primary-500  hover:border-ws-primary-500 peer grid h-7 w-7 shrink-0 appearance-none cursor-pointer place-content-center after:hidden after:content-[''] indeterminate:after:bg-indeterminate indeterminate:border-ws-primary-500 indeterminate:bg-ws-primary-500 indeterminate:after:block indeterminate:hover:border-ws-primary-700 indeterminate:hover:bg-ws-primary-700 indeterminate:after:h-[24px] indeterminate:after:w-[24px] checked:after:bg-checkmark checked:border-ws-primary-500 checked:bg-ws-primary-500 checked:after:block checked:hover:border-ws-primary-700 checked:hover:bg-ws-primary-700 checked:after:h-[13px] checked:after:w-[13px] disabled:border-ws-grey-100 disabled:bg-ws-grey-50 disabled:checked:border-ws-primary-300 disabled:checked:bg-ws-primary-300"
  id="default"
  type="checkbox"
/>
`;
